nsxt/resource_nsxt_upgrade_run.go

```diff
@@ -0,0 +1,718 @@
+		}
+		elemSchema["maintenance_mode_config_evacuate_powered_off_vms"] = &schema.Schema{
+			Type:         schema.TypeString,
+			Description:  "Maintenance mode config evacuate powered off vms",
+			Optional:     true,
+			ValidateFunc: validation.StringInSlice(supportedMaintenanceModeConfigEvacuateOffVms, false),
+		}
+		elemSchema["rebootless_upgrade"] = &schema.Schema{
+			Type:        schema.TypeBool,
+			Description: "Rebootless upgrade",
+			Optional:    true,
+			Default:     true,
+		}
+	}
+
+	return &schema.Schema{
+		Type:        schema.TypeList,
+		Description: "Upgrade group for this upgrade",
+		Optional:    true,
+		Elem: &schema.Resource{
+			Schema: elemSchema,
+		},
+	}
+}
+
+func getUpgradeSettingSchema(isEdge bool) *schema.Schema {
+	elemSchema := map[string]*schema.Schema{
+		"post_upgrade_check": {
+			Type:        schema.TypeBool,
+			Description: "Whether run post upgrade check",
+			Optional:    true,
+			Default:     true,
+		},
+		"parallel": {
+			Type:        schema.TypeBool,
+			Description: "Whether run upgrade parallel",
+			Optional:    true,
+			Default:     true,
+		},
+		"stop_on_error": {
+			Type:        schema.TypeBool,
+			Description: "Whether stop the upgrade when an error occur",
+			Optional:    true,
+			Default:     false,
+		},
+	}
+	if isEdge {
+		// Edge Upgrade setting is forced to stop on error.
+		delete(elemSchema, "stop_on_error")
+	}
+	return &schema.Schema{
+		Type:        schema.TypeList,
+		Description: "Upgrade plan setting for component",
+		Optional:    true,
+		MaxItems:    1,
+		Elem: &schema.Resource{
+			Schema: elemSchema,
+		},
+	}
+}
+
+func resourceNsxtUpgradeRunCreate(d *schema.ResourceData, m interface{}) error {
+	return upgradeRunCreateOrUpdate(d, m, true)
+}
+
+func upgradeRunCreateOrUpdate(d *schema.ResourceData, m interface{}, isCreate bool) error {
+	id := d.Id()
+	if id == "" {
+		id = newUUID()
+	}
+	connector := getPolicyConnectorWithHeaders(m, nil, false, false)
+	upgradeClientSet := newUpgradeClientSet(connector, d)
+
+	log.Printf("[INFO] Updating UpgradeUnitGroup and UpgradePlanSetting.")
+	err := prepareUpgrade(upgradeClientSet, d, isCreate)
+	if err != nil {
+		return handleCreateError("NsxtUpgradeRun", id, err)
+	}
+
+	log.Printf("[INFO] Successfully update UpgradeUnitGroup and UpgradePlanSetting. Start Upgrade.")
+
+	err = runUpgrade(upgradeClientSet, getPartialUpgradeMap(d))
+	if err != nil {
+		return handleCreateError("NsxtUpgradeRun", id, err)
+	}
+
+	runPostcheck(upgradeClientSet.UpgradeClient, d)
+
+	d.SetId(id)
+	return resourceNsxtUpgradeRunRead(d, m)
+}
+
+func prepareUpgrade(upgradeClientSet *upgradeClientSet, d *schema.ResourceData, isCreate bool) error {
+	for i := range upgradeComponentList {
+		component := upgradeComponentList[i]
+		// Customize MP upgrade is not allowed
+		if component == mpUpgradeGroup {
+			continue
+		}
+
+		if !isCreate && !d.HasChange(componentToGroupKey[component]) && !d.HasChange(componentToSettingKey[component]) {
```

Comment:
I think `HasChange` would be true for create use case - if so, looks like this parameter is not needed in this function

Reply:
Good catch. Done.